webknossos/webknossos/skeleton/entities.py

```diff
@@ -0,0 +1,479 @@
+    group_id: Optional[int] = None
+
+    _enforce_id: Optional[int] = None
+
+    def __attrs_post_init__(self) -> None:
+        self.nx_graph = nx.Graph()
+        if self._enforce_id is not None:
+            self.id = self._enforce_id
+        else:
+            self.id = self._nml.element_id_generator.__next__()
+
+    # todo: how to type this?
+    def get_nodes(self) -> Any:
+
+        return self.nx_graph.nodes
+
+    def get_node_positions(self) -> np.ndarray:
+        return np.array([node.position for node in self.nx_graph.nodes])
+
+    def get_node_by_id(self, node_id: int) -> Node:
+
+        # Todo: Use hashed access
+        for node in self.get_nodes():
+            if node.id == node_id:
+                return node
+        raise ValueError(f"No node with id {node_id} was found")
+
+    def add_node(
+        self,
+        position: Vector3,
+        comment: Optional[str] = None,
+        radius: Optional[float] = None,
+        rotation: Optional[Vector3] = None,
+        inVp: Optional[int] = None,
+        inMag: Optional[int] = None,
+        bitDepth: Optional[int] = None,
+        interpolation: Optional[bool] = None,
+        time: Optional[int] = None,
+        is_branchpoint: bool = False,
+        branchpoint_time: Optional[int] = None,
+        _enforce_id: Optional[int] = None,
+        _nml: Optional["NML"] = None,
+    ) -> Node:
+        node = Node(
+            position=position,
+            comment=comment,
+            radius=radius,
+            rotation=rotation,
+            inVp=inVp,
+            inMag=inMag,
+            bitDepth=bitDepth,
+            interpolation=interpolation,
+            time=time,
+            is_branchpoint=is_branchpoint,
+            branchpoint_time=branchpoint_time,
+            enforce_id=_enforce_id,
+            nml=_nml or self._nml,
+        )
+        self.nx_graph.add_node(node)
+        return node
+
+    def add_edge(self, node_1: Node, node_2: Node) -> None:
+
+        self.nx_graph.add_edge(node_1, node_2)
+
+    def get_max_node_id(self) -> int:
+
+        # Chain with [0] since max is not defined on empty sequences
+        return max(itertools.chain((node.id for node in self.nx_graph.nodes), [0]))
+
+    def __hash__(self) -> int:
+        return hash((self._nml.id, self.id))
+
+
+@attr.define()
+class NML:
+    """
+    Contains groups and skeletons.
+    """
+
+    name: str
+    id: int = attr.ib(init=False)
+    scale: Vector3
+    offset: Optional[Vector3] = None
+    time: Optional[int] = None
+    edit_position: Optional[Vector3] = None
+    edit_rotation: Optional[Vector3] = None
+    zoom_level: Optional[float] = None
+    task_bounding_box: Optional[IntVector6] = None
+    user_bounding_boxes: Optional[List[IntVector6]] = None
+
+    root_group: Group = attr.ib(init=False)
+    element_id_generator: Iterator[int] = attr.ib(init=False)
+
+    def __attrs_post_init__(self) -> None:
+        self.id = nml_id_generator.__next__()
```

Comment:
I'm not sure if we want to use a global counter, since this can lead to problems with threading, pickling, etc… Instead we could use `id(self)` where we had `id` before:
https://docs.python.org/3/library/functions.html#id

Reply:
remove counter